File: DemoServer/Indexes/ProductSales.cs
using System.Linq;
using DemoServer.Entities;
using Raven.Client.Indexes;

namespace DemoServer.Indexes
{
    public class ProductSales : AbstractIndexCreationTask<Order, ProductSales.Result>
    {
        public class Result
        {
            public string Product { get; set; }

            public int Count { get; set; }

            public decimal Total { get; set; }
        }

        public ProductSales()
        {
            Map = orders => from order in orders
                            from line in order.Lines
                            select new
                            {
                                Product = line.Product,
                                Count = 1,
                                Total = ((line.Quantity * line.PricePerUnit) * (1 - line.Discount))
                            };

            Reduce = results => from result in results
                                group result by result.Product into g
                                select new
                                {
                                    Product = g.Key,
                                    Count = g.Sum(x => x.Count),
                                    Total = g.Sum(x => x.Total)
                                };
        }
    }
}

File: DemoServer/JavaDemo/src/main/java/net/ravendb/demo/indexes/ProductSales.java
package net.ravendb.demo.indexes;

import net.ravendb.client.documents.indexes.AbstractIndexCreationTask;

public class ProductSales extends AbstractIndexCreationTask {

    public static class Result {
        private String product;
        private int count;
        private double total;

        public String getProduct() {
            return product;
        }

        public void setProduct(String product) {
            this.product = product;
        }

        public int getCount() {
            return count;
        }

        public void setCount(int count) {
            this.count = count;
        }

        public double getTotal() {
            return total;
        }

        public void setTotal(double total) {
            this.total = total;
        }
    }

    public ProductSales() {
        map = "from order in docs.Orders " +
                "from line in order.Lines " +
                "select new { " +
                "  Product = line.Product, " +
                "  Count = 1, " +
                "  Total = (((decimal)line.Quantity) * line.PricePerUnit) * (1M - line.Discount) " +
                "}";

        reduce = "from result in results " +
                "group result by result.Product into g " +
                "select new { " +
                "  Product = g.Key, " +
                "  Count = Enumerable.Sum(g, x => ((int)x.Count)), " +
                "  Total = Enumerable.Sum(g, x0 => ((decimal)x0.Total)) " +
                "}";
    }

}
